Issue/1921 extend version overview with more versions

Hugo-Inmanta opened the pull request:
# Description

Part of https://github.com/inmanta/irt/issues/1921



# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

Hugo-Inmanta commented:
Current look:
![Screenshot from 2024-04-11 17-10-08](https://github.com/inmanta/inmanta-core/assets/98876549/4088c2e5-4253-4be6-9475-678a1b8bfb43)
![Screenshot from 2024-04-11 17-10-25](https://github.com/inmanta/inmanta-core/assets/98876549/2a132725-4bf7-4c20-a1e6-d5f02e4ff6ff)

arnaudsjs commented on docs/_templates/module_sets.tmpl:
```diff
@@ -0,0 +1,21 @@
+
+   {% for key, value in data["module_set_constraints"] | dictsort %}
+   * - .. centered:: {{ key }}
+     - .. centered:: {{ value }}
```
How does this render if there is no version constraint?

Hugo-Inmanta replied:
The template engine can handle it well it seems:
```json
...
"module_set_constraints": {
        "lsm": "~=2.27",
        "std": null
    }
...
```
renders as: 
![image](https://github.com/inmanta/inmanta-core/assets/98876549/2da8b784-f7a0-4d86-87ff-0238ddfcd7aa)

Hugo-Inmanta commented on docs/reference/compatibility.rst:
```diff
@@ -1,20 +1,48 @@
+
+.. only:: oss
+
+    The table below shows the set of Inmanta modules that come along with this Inmanta OSS version |release|
```
Indeed, I will only list the modules for OSS :+1:

Hugo-Inmanta commented on docs/reference/compatibility.rst:
```diff
@@ -1,20 +1,48 @@
+.. only:: iso
+
+    The table below shows the set of Inmanta modules that come along with version |version_major| of the
+    Inmanta Service Orchestrator and constraints on their respective version (if any).
```
I'm not 100% sure about their meaning and the information we want to convey here, what about:
"These constraints ensure compatibility among the modules within this version of the Inmanta Service Orchestrator." ?

arnaudsjs replied:
I am not sure about this myself. Internally we use these constraints to make sure we don't release breaking changes within a major release. But I am not sure what information this represent for our customers. @bartv Do you know?

bartv replied:
I have no idea either why this is required. Same for all of the extensions but in core it seems to be something really important to add this to the docs. I can accept that but then it also needs to explanation so that the information has meaning or even worse that it does not cause any confusion.

arnaudsjs replied:
I agree. But nobody in the team seems to know why we want this in the docs. @wouterdb Do you?

wouterdb replied:
Two reasons why I think we need it for the modules:
1. if a module is developed for a specific ISO version (that is not the latest), these constraints allow you to keep your modules in-range even when only using a single distribution channel (which is the case for us internally). We use this a lot internally, but via the irt repo.  Being able to add these constraints also makes it less dangerous when the wrong distribution channel is configured by accident. 
2. if we get a support request, it allows us to verify that the modules involved are in-range for the orchestrator used. 

For the extensions: less common, we may choose to drop it. It is mostly important when using pytest-inmanta and lsm. Then you need an in-range version of inmanta-lsm but that can now be solved elegantly by installing inmanta-service-orchestrator~=X.0 .

Hugo-Inmanta commented:
Current looks:
![image](https://github.com/inmanta/inmanta-core/assets/98876549/54de1352-f60b-4d54-8221-0820116c2143)

associated json file:
```json
{
    "component_requirements": {
        "inmanta-core": "~=11.0.dev",
        "inmanta-license": "~=4.0.dev",
        "inmanta-lsm": "~=4.0.dev",
        "inmanta-support": "~=3.2.dev",
        "inmanta-ui": "~=5.1.dev"
    },
    "compiler_version": "2025.0",
    "system_requirements": {
        "python_version": "3.11",
        "rhel_versions": [
            9,
            8
        ],
        "postgres_version": 13
    },
    "module_compatibility_ranges": {
        "apt": "~=0.4",
        "aws": "~=4.0",
        "exec": "~=1.1",
        "ip": "~=2.0",
        "lsm": "~=2.27",
        "net": "~=1.1",
        "openstack": "~=4.0",
        "redhat": "~=1.0",
        "rest": "~=0.2",
        "ssh": "~=1.0",
        "std": "~=5.0",
        "ubuntu": "~=0.4",
        "user": "~=0.1",
        "yum": "~=0.7"
    }
}
```

arnaudsjs commented on docs/_templates/compiler_version.tmpl:
```diff
@@ -0,0 +1,24 @@
+Compiler version
+~~~~~~~~~~~~~~~~
+
+Specifying the compatible ISO version for a V1 module is done via the ``compiler_version`` field of the ``module.yml``
```
If there are no changes to the compiler, the compiler_version will remain the same across ISO versions. 
```suggestion
Specifying the compatible compiler version for a V1 module is done via the ``compiler_version`` field of the ``module.yml``
```

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
-   * - .. centered:: {{ key }}
-     - .. centered:: {{ value }}
-   {% endfor %}
+During module development, when the version of the module integrates tightly with a specific ISO version,
```
What do you mean with "when the version of the module integrates tightly with a specific ISO version". And why is that important?

After reading the remainder of this document, I think you want to say the following: "When an Inmanta module integrates tightly with the Inmanta Service Orchestrator, it's important to use a version of that module that is compatible with the orchestrator"

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
+we need a way to tell the CI which version of the module should be used per ISO line. This can be achieved by either:
 
-.. note::
+- Using a specific index per ISO version
```
I think I know what you mean here, namely that we have indexes/git repos that only contain the versions for a specific ISO version. But I think that an end-user will need a more elaborate explanation to be able to understand this.

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
 
-.. note::
+- Using a specific index per ISO version
+- Adding a project constraint file with the compatible module ranges for each ISO version.
```
Same for this point.

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
+
+.. only:: oss
+
+    The table below shows the set of Inmanta modules that integrate tightly with this Inmanta OSS version |release|
+    and the constraints on their version.
```
```suggestion
    The table below shows the set of Inmanta modules that integrate tightly with this Inmanta OSS version |release| and which versions are compatible with this OSS release.
```

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
+
+    .. note::
+
+        If for some reason you need to install any of the components that make up the inmanta-service-orchestrator
+        package separately e.g. when developing an Inmanta extension, the ranges of the Inmanta Python components this
+        product version is compatible with are available as a constraints file in
+        `requirements.components.txt <./requirements.components.txt>`_ and in json format in the
+        `compatibility.json <./compatibility.json>`_ file.
```
* This is one big sentence and I find it hard to read. Let's split this sentence into two or more sentences.
* I would first mention the normal procedure here. Namely installing the `inmanta-service-orchestrator~=X.0` to get all the components for a specific ISO version. And then mention that if you want to rely on a specific component(s) only, you need to take into account the following constraints.

arnaudsjs commented on docs/_templates/compiler_version.tmpl:
```diff
@@ -0,0 +1,24 @@
+
+.. only:: oss
+
+    Inmanta version |version| uses compiler version {{ data['compiler_version'] }}.
```
```suggestion
    Inmanta OSS version |version| uses compiler version {{ data['compiler_version'] }}.
```

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,76 @@
+.. only:: iso
+
+
+    The table below shows the set of Inmanta modules that integrate tightly with version |version_major| of the
+    Inmanta Service Orchestrator and the constraints on their version.
```
```suggestion
    The table below shows the set of Inmanta modules that integrate tightly with version |version_major| of the
    Inmanta Service Orchestrator and which versions are compatible with this ISO release.
```

arnaudsjs commented on docs/_templates/compiler_version.tmpl:
```diff
@@ -0,0 +1,24 @@
+Compiler version
+~~~~~~~~~~~~~~~~
+
+Specifying the compatible compiler version for a V1 module is done via the ``compiler_version`` field of the ``module.yml``
+file.
```
I just noticed that this field is nowhere explained in the documentation. Do you create a ticket to explain what the compiler_version field is about.

Hugo-Inmanta replied:
Decided to drop the compiler section as per https://inmanta.slack.com/archives/CKRF0C8R3/p1713441521052269?thread_ts=1713339151.435429&cid=CKRF0C8R3

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,83 @@
-.. note::
+- Using a specific index (or git repo) per ISO version and making sure it contains the compatible versions of the modules.
+- Using a generic index (or git repo) possibly containing multiple versions for a given module and adding a
+  project constraint file for each ISO version in which the compatible module ranges are specified.
```
What is a project constraint file?

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,83 @@
+to use a version of that module that is compatible with the orchestrator. This can be achieved by either:
 
-.. note::
+- Using a specific index (or git repo) per ISO version and making sure it contains the compatible versions of the modules.
+- Using a generic index (or git repo) possibly containing multiple versions for a given module and adding a
```
One pip index can host multiple modules, while there is only one git repo per module (V1). So repos has to be plural.
```suggestion
- Using a specific pip index (or git repos) per ISO version and making sure it only contains the compatible versions of the modules.
- Using a generic pip index (or git repos) containing module versions for different ISO versions and adding a
```

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,83 @@
+
+    .. note::
+
+        To transparently install the latest compatible versions for all the python components that make up the
+        ``inmanta-service-orchestrator`` package for version |version_major| of the Inmanta Service Orchestrator,
+        install it via the following command:
```
What do you mean with "transparantly". This method is not really transparent right, because all the version filtering happens in the background based on the version constraints embedded in the product package. Do you mean conveniently?
```suggestion
        To conveniently install the latest compatible versions for all the python components that make up version |version_major| of the Inmanta Service Orchestrator product,
        install the product package using the following command:
```

Hugo-Inmanta replied:
> What do you mean with "transparantly"

I meant it in the sense "without the user being aware" since it happens in the background without the need of specifying versions for components

arnaudsjs replied:
I see. I just learned that transparent can also have another meaning than the one I had in mind and that is what you are using it form.

Hugo-Inmanta replied:
I'll go with conveniently if it's clearer (pun intended :) )

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,83 @@
+
+.. only:: iso
+
+    {% if data['component_requirements'] %}
```
Don't we need a similar section for OSS?

Hugo-Inmanta commented:
Added a `python_package_constraints` section to the json that includes all python constraints : for components and v2 modules
Latest look: 
![image](https://github.com/inmanta/inmanta-core/assets/98876549/6ecc0e8d-86f5-4d1c-bdaf-a8cc58e9edf8)
Json data:
```json
{
    "python_package_constraints": {
        "inmanta-core": "~=11.0.dev",
        "inmanta-license": "~=4.0.dev",
        "inmanta-lsm": "~=4.0.dev",
        "inmanta-support": "~=3.2.dev",
        "inmanta-ui": "~=5.1.dev",
        "inmanta-module-apt": "~=0.4",
        "inmanta-module-aws": "~=4.0",
        "inmanta-module-exec": "~=1.1",
        "inmanta-module-ip": "~=2.0",
        "inmanta-module-lsm": "~=2.27",
        "inmanta-module-net": "~=1.1",
        "inmanta-module-openstack": "~=4.0",
        "inmanta-module-redhat": "~=1.0",
        "inmanta-module-rest": "~=0.2",
        "inmanta-module-ssh": "~=1.0",
        "inmanta-module-std": "~=5.0",
        "inmanta-module-ubuntu": "~=0.4",
        "inmanta-module-user": "~=0.1",
        "inmanta-module-yum": "~=0.7"
    },
    "system_requirements": {
        "python_version": "3.11",
        "rhel_versions": [
            9,
            8
        ],
        "postgres_version": 13
    },
    "module_compatibility_ranges": {
        "apt": "~=0.4",
        "aws": "~=4.0",
        "exec": "~=1.1",
        "ip": "~=2.0",
        "lsm": "~=2.27",
        "net": "~=1.1",
        "openstack": "~=4.0",
        "redhat": "~=1.0",
        "rest": "~=0.2",
        "ssh": "~=1.0",
        "std": "~=5.0",
        "ubuntu": "~=0.4",
        "user": "~=0.1",
        "yum": "~=0.7"
    }
}
```

Hugo-Inmanta commented:
@sanderr  requesting your review mostly for what we discussed:
- regrouping all python constraints in one section in the json file
- the rationale for the Module compatibility range section: any suggestion for a better wording or missing info is welcome

sanderr commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,116 @@
+
+.. note::
 
+    This information is also available in a machine-consumable format in the `compatibility.json <./compatibility.json>`_ file.
```
I would also mention the other modules, i.e. those that are not "most commonly used" as put above.

Hugo-Inmanta commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,116 @@
-.. note::
+.. only:: oss
+
+    The continuous integration can be configured to run tests against multiple OSS versions. Each version might require
+    a different set of Inmanta modules and different versions of these modules.
```
I tend to agree since we don't specify a compatible range for oss modules anyways

Hugo-Inmanta commented:
> The files look good to me, though I would prefer to change the name of the txt file ("components" isn't really accurate anymore). But that may mess up some builds?

As far as I know it is used in 3 builds from core but I think it can be changed. Do you have a new name suggestion ? `"requirements.python_constraints.txt"` ?

sanderr commented:
I think that's both very verbose and redundant (it is by definition a python constraints file). Perhaps `requirements.internal.txt`? I don't like it so much, but it's the best I can come up with. Or perhaps we can consider plain `requirements.txt`. It does after all constrain everything that we consider part of an iso line now.

I think I'm in favor of the latter. But I'd check if there are other opinions.

sanderr commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,108 @@
+        .. code-block:: sh
+            :substitutions:
+
+            pip install inmanta~=|release|
```
I would even go so far as to drop the constraint entirely. And now that I think about it, I think we might as well drop the entire section for OSS, just like we did the module constraints section. The same arguments apply here.

Hugo-Inmanta replied:
Dropped the section

sanderr commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,108 @@
 
-    This information is also available as a constraints file in `requirements.components.txt <./requirements.components.txt>`_
-    and in json format in the `compatibility.json <./compatibility.json>`_ file.
+    For stability reasons, when you depend on these modules, we encourage you to ensure that you use the appropriate
+    versions for the iso version you're using. This can be achieved one of two ways:
```
I don't think we've ever discussed that. I am not opposed per se, and perhaps you're right that we should enforce it, but in practice other versions might work just as well.

Hugo-Inmanta replied:
Went with: 
```
For stability reasons, when you depend on these modules, make sure you use the appropriate
versions for the iso version you're using. This can be achieved one of two ways:
```

sanderr reviewed (changes requested):
Did you forget about the file rename?

arnaudsjs commented on docs/_templates/components_requirements.tmpl:
```diff
@@ -1,39 +1,108 @@
-   * - .. centered:: {{ key }}
-     - .. centered:: {{ value }}
-   {% endfor %}
+    We develop and distribute some core modules alongside the product. For these modules, we uphold the same
+    compatibility guarantees as for the product itself. Therefore, some versions of a module may be developed
+    for a specific version of the product (e.g. a breaking change in a module is intended only for the next major
+    iso release).
```
To me it feel like this sentence and the previous one say exactly the same with different words. Together with the example, it feel very repetitive and verbose to me. But I don't have any objections against keeping it like it was.

Hugo-Inmanta replied:
I find the verbose form easier to understand

arnaudsjs replied:
Let's keep it then.

Hugo-Inmanta commented:
> Did you forget about the file rename?

At the time of the PR I hadn't renamed the `INMANTA_REQUIREMENTS_COMPONENTS_TXT` -> `INMANTA_REQUIREMENTS_TXT` env variable in irt.

Now that it got renamed I updated it in the `tox.ini` file

And I also renamed it in the last section of the `components_requirements.tmpl` file.

Is there another location you have in mind that I'm forgetting about ?

sanderr reviewed (approved):
To be clear, I've only reviewed the parts you explicitly asked me about (and some few things that happened to catch my eye).

inmantaci commented:
Processing this pull request

inmantaci commented:
Merged into branches master in 59ca8186013f8200bc7b18dbdfdfd229df0e6ca4

inmantaci commented:
Processing #7568.

inmantaci commented:
Processing #7567.
